### DISCUSSION TITLE
Troubles with Swarm Demo

### DISCUSSION THREAD

**Discussion by CRhodes-123** [bitcraze/discussions#326] [Q&A]
Hello everyone,

I am Collin Rhodes, a University student doing research with crazyflies.

I have recently set up a lighthouse positioning system, and I'm super happy to have gotten that set up. It has two V2 base stations, and flying the drone with the controller in the crazyflie client is really cool.

I am now trying to get the swarm demos to work, and the one I am trying right now is the hl-commander-swarm.py demo. For reference, I am using three crazyflies with the lighthouse and flow decks attached, and each crazyflie has the newest firmware release and the calibration and geometry data for the lighthouse base stations in permanent memory. Whenever I try to run the demo, I get this error:
![image](https://user-images.githubusercontent.com/103008098/196548541-cee1ca7e-3179-4845-8176-0384e713e4cd.png)

I will reply to this post with the full code exactly as I am running it for your convenience.

Any help in identifying what this means/what error I need to correct would be greatly appreciated.

Thanks, 
Collin Rhodes

**Comment by CRhodes-123** [comment]
Here is the code exactly as I am running it:
``` python
# -*- coding: utf-8 -*-
#
#     ||          ____  _ __
#  +------+      / __ )(_) /_______________ _____  ___
#  | 0xBC |     / __  / / __/ ___/ ___/ __ `/_  / / _ \
#  +------+    / /_/ / / /_/ /__/ /  / /_/ / / /_/  __/
#   ||  ||    /_____/_/\__/\___/_/   \__,_/ /___/\___/
#
#  Copyright (C) 2019 Bitcraze AB
#
#  This program is free software; you can redistribute it and/or
#  modify it under the terms of the GNU General Public License
#  as published by the Free Software Foundation; either version 2
#  of the License, or (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <https://www.gnu.org/licenses/>.
"""
Simple example of a swarm using the High level commander.

The swarm takes off and flies a synchronous square shape before landing.
The trajectories are relative to the starting positions and the Crazyflies can
be at any position on the floor when the script is started.

This example is intended to work with any absolute positioning system.
It aims at documenting how to use the High Level Commander together with
the Swarm class.
"""
import time

import cflib.crtp
from cflib.crazyflie.swarm import CachedCfFactory
from cflib.crazyflie.swarm import Swarm


def activate_high_level_commander(scf):
    scf.cf.param.set_value('commander.enHighLevel', '1')


def activate_mellinger_controller(scf, use_mellinger):
    controller = 1
    if use_mellinger:
        controller = 2
    scf.cf.param.set_value('stabilizer.controller', controller)


def run_shared_sequence(scf):
    activate_mellinger_controller(scf, False)

    box_size = 1
    flight_time = 2

    commander = scf.cf.high_level_commander

    commander.takeoff(1.0, 2.0)
    time.sleep(3)

    commander.go_to(box_size, 0, 0, 0, flight_time, relative=True)
    time.sleep(flight_time)

    commander.go_to(0, box_size, 0, 0, flight_time, relative=True)
    time.sleep(flight_time)

    commander.go_to(-box_size, 0, 0, 0, flight_time, relative=True)
    time.sleep(flight_time)

    commander.go_to(0, -box_size, 0, 0, flight_time, relative=True)
    time.sleep(flight_time)

    commander.land(0.0, 2.0)
    time.sleep(2)

    commander.stop()


uris = {
    'radio://0/30/2M/E7E7E7E711',
    'radio://0/30/2M/E7E7E7E712',
    # Add more URIs if you want more copters in the swarm
}

if __name__ == '__main__':
    cflib.crtp.init_drivers()
    factory = CachedCfFactory(rw_cache='./cache')
    with Swarm(uris, factory=factory) as swarm:
        swarm.parallel_safe(activate_high_level_commander)
        swarm.reset_estimators()
        swarm.parallel_safe(run_shared_sequence)
```

**Reply by krichardsson** [reply]
It is hard to tell what the problem is, the underlying exception that caused the problem is not visible. You can probably find it in your teminal output in vscode.

The most likely cause is that it did not manage to connect to the crazyflie(s). Are your Crazyflies set to the addresses in the uri list?
``` python
uris = {
    'radio://0/30/2M/E7E7E7E711',
    'radio://0/30/2M/E7E7E7E712',
    # Add more URIs if you want more copters in the swarm
}
```

Can you connect to the Crazyflies with the client, using the URIs?

**Reply by CRhodes-123** [reply]
I haven’t really been able to connect to all of the crazyflies at once in the crazyflie client yet. Am I supposed to use multiple crazy radios to do this? If not, how do I figure out which uri to use?

**Reply by CRhodes-123** [reply]
In the lab today, I got this when trying to run the demo when I had added another uri to account for the 3rd crazyflie:
![image](https://user-images.githubusercontent.com/103008098/196816932-b2d042da-f184-40a7-9f35-cd920d896bfa.png)

I then tried to follow this youtube video to see how to run the script properly and if there were any errors that I was making: https://youtu.be/LG7_gyhrPqE
At about 4 minutes, it describes how to flash multiple crazyflies at once, and the make part of the script worked correctly, but when i tried to run cload-all.sh, this is what I got:
![image](https://user-images.githubusercontent.com/103008098/196817143-f8f69170-8d92-4917-b4c2-21ca814f29a1.png)

For reference, this is what I see when i connect to one crazyflie in the client:
![image](https://user-images.githubusercontent.com/103008098/196817190-db74b551-421d-412b-bc3a-3b2162205cbd.png)

Hopefully this is helpful in diagnosing the problem.

Thanks,
Collin Rhodes

**Reply by CRhodes-123** [reply]
Also, is there a way to connect to multiple crazyflies at once in the crazyflie client? 
![image](https://user-images.githubusercontent.com/103008098/196818292-8ad9bede-9b6f-4d20-ac9f-49480f266262.png)
My understanding is that the app that launches in the virtual machine when i double click on this icon is the crazyflie client, but I do not see any options for connecting to multiple crazyflies, and when I change the address and only one crazyflie is turned on, they connect without any issues for all of the addresses I have tried (0xE7E7E7E7E1 - 0xE7E7E7E7E9). This is the case for all of the crazyflies

Thanks,
Collin Rhodes

**Reply by CRhodes-123** [reply]
Today I changed the uri's manually in the script, and when I added an extra E7 at the end of the address I got this exception, saying the unpack requires a buffer of 5 bytes:
![image](https://user-images.githubusercontent.com/103008098/197024155-fdace8e2-68aa-449e-a157-77c35718f76b.png)

**Comment by krichardsson** [comment]
## URIs

First of all, we need to understand the URI. (I can see that we are missing documentation for this)
The URI describes the connection link to the Crazyflie and tells a client where to find it. The two most common ones are for
### Crazyradio 

For instance `radio://0/20/2M/E7E7E7E704`

`radio://<index>/<channel>/<bitrate>/<address>`

* `radio://`: use a Crazyradio
* index: 0, 1, 2 and so on for each Crazyradio connected to USB on the computer
* channel: The radio channel (frequency) 0 - 120
* bitrate: 250K, 1M or 2M. The bitrate on the link, usually use 2M
* address: a 10 digit hexadecimal number. If no address is stated the default address is E7E7E7E7E7. (`radio://0/20/2M` == `radio://0/20/2M/E7E7E7E7E7`)

Two Crazyflies must not use the same channel and address pair. If they use the same channel they must have different addresses or vice versa. 

Two radios should not use the same channel.

The most efficient setup is to use the same channel for all Crazyflies connected to the same radio. An example setup for two radios could be:
 ```
radio://0/20/2M/E7E7E7E701
radio://0/20/2M/E7E7E7E702
radio://1/80/2M/E7E7E7E701
radio://1/80/2M/E7E7E7E702
```

where radio 0 uses channel 20 and radio 1 uses channel 80.

Instructions on [how to set the channel/address of a crazyflie](https://www.bitcraze.io/documentation/repository/crazyflie-clients-python/master/userguides/userguide_client/#firmware-configuration)

### USB (cable)

If a crazyflie is connected directly to the computer via a USB cable, the URI is `USB://0`

## Client

The Client can only connect to one Crazyflie, there is no support for multiple connections.

## What to do

First configure your Crazyflies to use the same channel but different addresses. Use the client as described in the link above. Remember to restart the Crazflie after you change channel/address as it only changed after reboot.

Secondly change the URIs in the script to match what you set in Crazyflies. 

I think this will solve your problems.

If you want to connect to a crazyflie from your client, for instance `radio://1/80/2M/E7E7E7E701`. First enter the address `E7E7E7E701` in the address box and click scan. Now the Crazyflie should be found and `radio://1/80/2M/E7E7E7E701` be available in the URI dropdown box. It is only possible to scan for a Crazyflie if you know the address.

If you have forgotten the address of a Crazyflie you can always connect via USB to find/change the address from the client.

**Reply by kulasdrew** [reply]
I changed the address under configure 2.X to E7E7E7E701 and now when I try to connect to the Crazyflie in the client it posts an error saying:
Failed to connect on radio://0/80/2M: Too many packets lost in a pop-up window within the client.  

What is a work around to this?  I am able to connect to other crazyflies under the E7E7E7E7E7 address still.

**Reply by krichardsson** [reply]
Please start a new discussion as this is a new question that does fit the title of the thread.
Thanks!
